grommet/grommet
# [TS] Migrate Accordion tests to .tsx

#### What does this PR do?

This PR migrates Accordion component tests from `*.js` to `*.tsx` to allow `ts-jest` to catch type-related errors in the unit testing pipeline. A few changes had to be made to support this, and are justified below.

#### Where should the reviewer start?

- Ensure all tests are currently passing
- Refer to the background context section below for more information about the initial errors picked up by TypeScript and the fixed implemented in this PR.

#### What testing has been done on this PR?

- Manually ran and passed Accordion component tests locally.
- CircleCI tests have passed.

#### How should this be manually tested?

- The reviewer should verify all changes made to tests and ensure that:
    - All test removals are due to unused legacy code and will not cause regressions
    - Modified tests behave identically as before and modifications are type-related.

#### Any background context you want to provide?

Upon initial renaming of tests, the following errors were picked up by `ts-jest`:

**Input**: `yarn test Accordion`
**Output**:
```
yarn run v1.22.10
$ jest --runInBand Accordion
 FAIL  src/js/components/Accordion/__tests__/Accordion-test.tsx
  ● Test suite failed to run

    src/js/components/Accordion/__tests__/Accordion-test.tsx:9:57 - error TS7016: Could not find a declaration file for module '../..'. 'grommet/src/js/components/index.js' implicitly has an 'any' type.

    9 import { Accordion, AccordionPanel, Box, Grommet } from '../..';
                                                              ~~~~~~~
    src/js/components/Accordion/__tests__/Accordion-test.tsx:342:22 - error TS7031: Binding element 'index' implicitly has an 'any' type.

    342     const Panel = ({ index }) => (
                             ~~~~~

Jest: "global" coverage threshold for statements (86.9%) not met: 0%
Jest: "global" coverage threshold for branches (76.3%) not met: 0%
Jest: "global" coverage threshold for lines (87.9%) not met: 0%
Jest: "global" coverage threshold for functions (88%) not met: 0%
Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        15.841 s
Ran all test suites matching /Accordion/i.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

- The first error was fixed by copying the contents of `grommet/src/js/components/index.js` to a new file `grommet/src/js/components/index.d.ts`, making types available for the typescript compiler. We do this for `index.d.ts` files elsewhere in the repo, it seems wew just missed this one.
- The second error is easily fixed by adding typings for the `index` prop to the `Panel` component. The `noImplicitAny` TypeScript rule forces us to provide explicit types where they otherwise couldn't be inferred (an implicit 'any' type). In the context of this unit test, `index` is always a number, so I added the appropriate typings to the test.

#### What are the relevant issues?

Parent tracking issue: #5599

#### Do the grommet docs need to be updated?

No

#### Should this PR be mentioned in the release notes?

No, all fixes are test-related and not component-related.

#### Is this change backwards compatible or is it a breaking change?

Backwards compatible.
